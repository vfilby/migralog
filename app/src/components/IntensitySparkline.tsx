import React from 'react';
import { View, StyleSheet } from 'react-native';
import Svg, { Path, Circle, Defs, LinearGradient, Stop, Rect } from 'react-native-svg';
import { getPainColor, PAIN_SCALE } from '../utils/painScale';

interface IntensitySparklineProps {
  /**
   * Array of intensity readings (0-10 scale)
   */
  intensities: number[];
  /**
   * Width of the sparkline in pixels
   */
  width?: number;
  /**
   * Height of the sparkline in pixels
   */
  height?: number;
  /**
   * Color of the line (defaults to peak intensity color)
   */
  color?: string;
  /**
   * Highlight the peak intensity point
   */
  showPeak?: boolean;
}

const IntensitySparkline: React.FC<IntensitySparklineProps> = ({
  intensities,
  width = 120,
  height = 40,
  color,
  showPeak = true,
}) => {
  if (!intensities || intensities.length === 0) {
    return null;
  }

  // Ensure we have valid data
  const validIntensities = intensities.filter(i => typeof i === 'number' && !isNaN(i));
  if (validIntensities.length === 0) {
    return null;
  }

  // Find min/max for scaling
  const minIntensity = 0; // Always start from 0 for pain scale
  const maxIntensity = 10; // Max pain scale
  const peakIntensity = Math.max(...validIntensities);
  const peakIndex = validIntensities.indexOf(peakIntensity);

  // Calculate path
  const padding = 4;
  const chartWidth = width - (padding * 2);
  const chartHeight = height - (padding * 2);
  const xStep = chartWidth / (validIntensities.length - 1 || 1);

  // Apply Exponential Moving Average (EMA) for smoothing
  const alpha = 0.75; // Smoothing factor (0 < alpha < 1, lower = smoother, higher = more responsive)
  const smoothedIntensities = validIntensities.reduce<number[]>((acc, intensity, index) => {
    if (index === 0) {
      acc.push(intensity);
    } else {
      // EMA formula: EMA(t) = alpha * value(t) + (1 - alpha) * EMA(t-1)
      const ema = alpha * intensity + (1 - alpha) * acc[index - 1];
      acc.push(ema);
    }
    return acc;
  }, []);

  // Generate smooth path coordinates using quadratic curves
  const pathData = smoothedIntensities
    .map((intensity, index) => {
      const x = padding + (index * xStep);
      const normalizedY = (intensity - minIntensity) / (maxIntensity - minIntensity);
      // Invert Y coordinate (SVG Y increases downward)
      const y = padding + chartHeight - (normalizedY * chartHeight);

      if (index === 0) {
        return `M ${x},${y}`;
      } else {
        // Use quadratic bezier curve for smooth interpolation
        const prevX = padding + ((index - 1) * xStep);
        const prevIntensity = smoothedIntensities[index - 1];
        const prevNormalizedY = (prevIntensity - minIntensity) / (maxIntensity - minIntensity);
        const prevY = padding + chartHeight - (prevNormalizedY * chartHeight);

        // Control point is midway between points
        const cpX = (prevX + x) / 2;

        return `Q ${cpX},${prevY} ${x},${y}`;
      }
    })
    .join(' ');

  // Peak point coordinates
  const peakX = padding + (peakIndex * xStep);
  const peakNormalizedY = (peakIntensity - minIntensity) / (maxIntensity - minIntensity);
  const peakY = padding + chartHeight - (peakNormalizedY * chartHeight);

  const lineColor = color || getPainColor(peakIntensity);

  return (
    <View style={styles.container}>
      <Svg width={width} height={height}>
        <Defs>
          {/* Pain scale gradient - bottom (green) to top (purple) */}
          <LinearGradient id="painGradient" x1="0" y1="1" x2="0" y2="0">
            {/* Create gradient stops for each pain level */}
            {PAIN_SCALE.map((level, index) => (
              <Stop
                key={level.value}
                offset={`${(index / (PAIN_SCALE.length - 1)) * 100}%`}
                stopColor={level.color}
                stopOpacity="0.15"
              />
            ))}
          </LinearGradient>
        </Defs>

        {/* Background gradient showing pain scale */}
        <Rect
          x={0}
          y={0}
          width={width}
          height={height}
          fill="url(#painGradient)"
          rx={4}
        />

        {/* Sparkline path */}
        <Path
          d={pathData}
          stroke={lineColor}
          strokeWidth={2}
          fill="none"
          strokeLinecap="round"
          strokeLinejoin="round"
        />

        {/* Peak intensity marker */}
        {showPeak && (
          <Circle
            cx={peakX}
            cy={peakY}
            r={3}
            fill={lineColor}
            stroke="white"
            strokeWidth={1.5}
          />
        )}
      </Svg>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
  },
});

export default IntensitySparkline;
