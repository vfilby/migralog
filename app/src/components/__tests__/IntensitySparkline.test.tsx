import React from 'react';
import { render } from '@testing-library/react-native';
import IntensitySparkline from '../IntensitySparkline';

describe('IntensitySparkline', () => {
  describe('Rendering', () => {
    it('should render without crashing with valid data', () => {
      const intensities = [3, 5, 7, 6, 4];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={120} height={40} />
      );
      expect(toJSON()).toBeTruthy();
    });

    it('should render null when intensities array is empty', () => {
      const { toJSON } = render(
        <IntensitySparkline intensities={[]} width={120} height={40} />
      );
      expect(toJSON()).toBeNull();
    });

    it('should render null when intensities is undefined', () => {
      const { toJSON } = render(
        <IntensitySparkline intensities={undefined as any} width={120} height={40} />
      );
      expect(toJSON()).toBeNull();
    });

    it('should handle array with single intensity value', () => {
      const intensities = [5];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={120} height={40} />
      );
      expect(toJSON()).toBeTruthy();
    });

    it('should filter out invalid intensity values', () => {
      const intensities = [3, NaN, 5, undefined as any, 7, null as any, 6];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={120} height={40} />
      );
      // Should render with only valid values [3, 5, 7, 6]
      expect(toJSON()).toBeTruthy();
    });

    it('should render null when all intensity values are invalid', () => {
      const intensities = [NaN, undefined as any, null as any];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={120} height={40} />
      );
      expect(toJSON()).toBeNull();
    });
  });

  describe('Props', () => {
    it('should use default width and height when not provided', () => {
      const intensities = [3, 5, 7];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} />
      );
      expect(toJSON()).toBeTruthy();
    });

    it('should accept custom width and height', () => {
      const intensities = [3, 5, 7];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={200} height={60} />
      );
      expect(toJSON()).toBeTruthy();
    });

    it('should accept custom color', () => {
      const intensities = [3, 5, 7];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} color="#FF0000" />
      );
      expect(toJSON()).toBeTruthy();
    });

    it('should hide peak marker when showPeak is false', () => {
      const intensities = [3, 5, 7, 6, 4];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} showPeak={false} />
      );
      expect(toJSON()).toBeTruthy();
    });

    it('should show peak marker by default', () => {
      const intensities = [3, 5, 7, 6, 4];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} />
      );
      expect(toJSON()).toBeTruthy();
    });
  });

  describe('Data Handling', () => {
    it('should handle intensity values at min boundary (0)', () => {
      const intensities = [0, 2, 4, 3, 1];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={120} height={40} />
      );
      expect(toJSON()).toBeTruthy();
    });

    it('should handle intensity values at max boundary (10)', () => {
      const intensities = [8, 9, 10, 9, 8];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={120} height={40} />
      );
      expect(toJSON()).toBeTruthy();
    });

    it('should handle full range of intensity values (0-10)', () => {
      const intensities = [0, 2, 4, 6, 8, 10, 7, 5, 3, 1];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={120} height={40} />
      );
      expect(toJSON()).toBeTruthy();
    });

    it('should handle all same intensity values', () => {
      const intensities = [5, 5, 5, 5, 5];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={120} height={40} />
      );
      expect(toJSON()).toBeTruthy();
    });

    it('should handle large number of intensity readings', () => {
      const intensities = Array.from({ length: 100 }, (_, i) =>
        Math.floor(Math.random() * 11)
      );
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={120} height={40} />
      );
      expect(toJSON()).toBeTruthy();
    });

    it('should handle ascending intensity pattern', () => {
      const intensities = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={120} height={40} />
      );
      expect(toJSON()).toBeTruthy();
    });

    it('should handle descending intensity pattern', () => {
      const intensities = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={120} height={40} />
      );
      expect(toJSON()).toBeTruthy();
    });

    it('should handle peak-then-recovery pattern', () => {
      const intensities = [2, 4, 6, 8, 10, 8, 6, 4, 2];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={120} height={40} />
      );
      expect(toJSON()).toBeTruthy();
    });
  });

  describe('Edge Cases', () => {
    it('should handle two intensity readings', () => {
      const intensities = [3, 7];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={120} height={40} />
      );
      expect(toJSON()).toBeTruthy();
    });

    it('should handle negative intensity values by filtering them out', () => {
      const intensities = [-1, 3, 5, -2, 7];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={120} height={40} />
      );
      // Should render with filtered positive values
      expect(toJSON()).toBeTruthy();
    });

    it('should handle intensity values above 10 by including them', () => {
      const intensities = [5, 7, 12, 6];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={120} height={40} />
      );
      // Should still render (scale adjusts to data)
      expect(toJSON()).toBeTruthy();
    });

    it('should handle very small dimensions', () => {
      const intensities = [3, 5, 7];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={20} height={10} />
      );
      expect(toJSON()).toBeTruthy();
    });

    it('should handle very large dimensions', () => {
      const intensities = [3, 5, 7];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={500} height={200} />
      );
      expect(toJSON()).toBeTruthy();
    });
  });

  describe('Snapshot Tests', () => {
    it('should match snapshot with typical intensity pattern', () => {
      const intensities = [3, 5, 7, 9, 8, 6, 4];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} width={120} height={40} />
      );
      expect(toJSON()).toMatchSnapshot();
    });

    it('should match snapshot with custom color', () => {
      const intensities = [3, 5, 7];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} color="#FF6B6B" width={120} height={40} />
      );
      expect(toJSON()).toMatchSnapshot();
    });

    it('should match snapshot without peak marker', () => {
      const intensities = [3, 5, 7, 6, 4];
      const { toJSON } = render(
        <IntensitySparkline intensities={intensities} showPeak={false} width={120} height={40} />
      );
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
