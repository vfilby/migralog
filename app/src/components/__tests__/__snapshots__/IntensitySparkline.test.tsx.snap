// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`IntensitySparkline Snapshot Tests should match snapshot with custom color 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
>
  <RNSVGSvgView
    bbHeight={40}
    bbWidth={120}
    focusable={false}
    height={40}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "flex": 0,
          "height": 40,
          "width": 120,
        },
      ]
    }
    width={120}
  >
    <RNSVGGroup
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
    >
      <RNSVGPath
        d="M 4,26.4 Q 32,26.4 60,21.6 Q 88,21.6 116,15.600000000000001"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 4294929259,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGCircle
        cx={116}
        cy={13.600000000000001}
        fill={
          {
            "payload": 4294929259,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        r={3}
        stroke={
          {
            "payload": 4294967295,
            "type": 0,
          }
        }
        strokeWidth={1.5}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`IntensitySparkline Snapshot Tests should match snapshot with typical intensity pattern 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
>
  <RNSVGSvgView
    bbHeight={40}
    bbWidth={120}
    focusable={false}
    height={40}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "flex": 0,
          "height": 40,
          "width": 120,
        },
      ]
    }
    width={120}
  >
    <RNSVGGroup
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
    >
      <RNSVGPath
        d="M 4,26.4 Q 13.333333333333334,26.4 22.666666666666668,21.6 Q 32,21.6 41.333333333333336,15.600000000000001 Q 50.66666666666667,15.600000000000001 60,9.3 Q 69.33333333333334,9.3 78.66666666666667,10.125 Q 88,10.125 97.33333333333334,15.131250000000001 Q 106.66666666666667,15.131250000000001 116,21.1828125"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 4293673082,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGCircle
        cx={60}
        cy={7.199999999999999}
        fill={
          {
            "payload": 4293673082,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        r={3}
        stroke={
          {
            "payload": 4294967295,
            "type": 0,
          }
        }
        strokeWidth={1.5}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`IntensitySparkline Snapshot Tests should match snapshot without peak marker 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
>
  <RNSVGSvgView
    bbHeight={40}
    bbWidth={120}
    focusable={false}
    height={40}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "flex": 0,
          "height": 40,
          "width": 120,
        },
      ]
    }
    width={120}
  >
    <RNSVGGroup
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
    >
      <RNSVGPath
        d="M 4,26.4 Q 18,26.4 32,21.6 Q 46,21.6 60,15.600000000000001 Q 74,15.600000000000001 88,16.5 Q 102,16.5 116,21.525"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 4292363029,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
